Room/app/src/main/java/com/example/room/FriendViewModel.kt

```diff
@@ -0,0 +1,132 @@
+package com.example.room
+
+import android.util.Patterns
+import androidx.lifecycle.*
+import com.example.room.db.Friend
+import com.example.room.db.FriendRepository
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+
+class FriendViewModel(private val friendRepository: FriendRepository) : ViewModel() {
+
+    fun getSaveFriends() = liveData {
+        friendRepository.friends.collect {
+            emit(it)
+        }
+    }
+
+    private var isUpdateOrDelete = false
+    private lateinit var friendToUpdateOrDelete: Friend
+
+    val inputName = MutableLiveData<String?>()
+    val inputEmail = MutableLiveData<String?>()
+
+    var saveOrUpdateButtonText = MutableLiveData<String>()
+    val deleteAllOrDeleteButtonText = MutableLiveData<String>()
+
+    private val statusMessage = MutableLiveData<Event<String>>()
+
+    val message: LiveData<Event<String>> get() = statusMessage
+
+    init {
+        saveOrUpdateButtonText.value = "저장"
+        deleteAllOrDeleteButtonText.value = "모두 삭제"
+    }
+
+    fun saveOrUpdate() {
+
+        if (inputName.value == null) {
+            statusMessage.value = Event("친구 이름을 입력해주세요.")
+        } else if (inputEmail.value == null) {
+            statusMessage.value = Event("친구 이메일을 입력해주세요.")
+        } else if (!Patterns.EMAIL_ADDRESS.matcher(inputEmail.value!!).matches()) {
+            statusMessage.value = Event("이메일 형식이 잘못됐습니다.")
+        } else {
+            if (isUpdateOrDelete) {
+                friendToUpdateOrDelete.name = inputName.value!!
+                friendToUpdateOrDelete.email = inputEmail.value!!
+                updateFriend(friendToUpdateOrDelete)
+            } else {
+                val name = inputName.value!!
+                val email = inputEmail.value!!
+                insertFriend(Friend(0, name, email))
+                inputName.value = null
+                inputEmail.value = null
+            }
+        }
+    }
+
+    fun deleteAllOrDelete() {
+        if (isUpdateOrDelete) {
+            deleteFriend(friendToUpdateOrDelete)
+        } else {
+            deleteAll()
+        }
+    }
+
+    fun insertFriend(friend: Friend) {
+        viewModelScope.launch {
```

Comment:
```suggestion
        viewModelScope.launch(Dispatchers.IO) {
```

Dispatchers를 지정해주지 않으면 default Dispatchers인 메인스레드에서 실행됩니다!
네트워크, 디스크, DB 작업에 적합한 Dispatchers.IO로 지정하는 것이 좋겠습니다 😎

Reply:
아하 주의하겠습니다!!!